# Watershed


Watershed is an unsupervised probabilistic framework that integrates genomic annotations and RNA-seq outlier calls into one model to identify the probability a rare variant has a functional effect on a particular RNA-seq outlier phenotype (examples of phenotypes can be, but are not limited to total expression, splicing, or ASE). Watershed extends our previous model RIVER (which can also be run from this package) by incoorperating information from multiple outlier phenoytpes into one model, where predictions for functional effects in one outlier phenotype are informed by observed outlier calls in another phenotype. Please see [ref bioRxiv preprint] for more details.


## Input file
Watershed models instances of gene-individual pairs. Therefor each line in the Watershed input file must be a gene-individual pair. Each line must include a minimum of one genomic annotations describing the rare variants nearby the gene for that individual. Each line must also include a minimum of one outlier score (p-value) describing the gene-individual pair. To evaluate Watershed performance we rely on N2 pairs, or pairs of individuals that share the same rare genetic variant nearby a particular gene (see bioRxiv preprint for more details). Therefor, each line in the input file must contain information concerning whether the gene-individual is an N2 pair. If the gene-individual is not an N2 pair, it can be represented as "NA". If the gene-individual is in an N2 pair, it can be represented as a positive integer that represents the unique identifier for that N2 pair. Note: if there are more than two individauls with the same rare genetic variant, we randomly select two individuals and ignore all other individuals. 

More formally, assuming G genomic annotations and E outlier p-values, the columns of the Watershed input file are as follows (column ordering is essential):
1. "SubjectID": Identifier for the individual in the gene-individual pair
2. "GeneName": Identifier for the gene in the gene-individual pair
3. A column for each of the G genomic annotations. Header names of these columns can be entirely user specified. Values are real valued.
4. A column for each of the E outlier p-values. Header names of these columns can be entirely user specified. Values are p-values (ie [0,1]). 
5. "N2pair": Identifier for whether gene-individual is in an N2 pair. If it is not, use "NA". If it is an N2 pair, use a unique positive integer to identify the pair.

An example input file with 18 genomic annotations and 3 outlier p-values can be found in "example_data/watershed_example_data.txt".
Another example input file with 18 genomic annotations and 1 outlier p-value can be found in "example_data/river_example_data_pheno_1.txt"

Note on outlier p-values: We can explicitely model outlier calls with sign (ie. total expression outliers are generated from a zscore and can therefor be over-expression or under-expression). To model this, if a particular outlier signal has sign/direction, simply put a negative sign in front of the p-value for instances that are under-expression. An exmample of this can be seen in the "pheno_2_pvalue" column of "example_data/watershed_example_data.txt".


## Running Watershed
There are two main scripts useful to user looking to apply Watershed to their data:
1. "evaluate_watershed.R": Briefly, this script will train a Watershed model on non-N2 pairs, and evaluate (via precision recall curves) on held-out N2 pairs. This will allow the user to get an idea of the accuracy of Watershed applied to their data.
2. "predict_watershed.R": Briefly, this script allows a user to train a Watershed model on training data, and then predict Watershed posterior probabilities (using Watershed parameters optimized in training) on all gene-individual in a much larger prediction data set.


## Running evaluate_watershed.R
evaluate_watershed.R can be easily run by using the following command:

Rscript evaluate_watershed.R --input $input_file --number_dimensions $number_of_dimensions --output_prefix $output_prefix --model_name $model 
where:
1. $input_file is the filename of the Watershed input file
2. $number_of_dimensions is an integer representing the number of outlier types (variable E in "input file" section)
3. $output_prefix is the prefix of all output files generated by evaluate_watershed.R
4. $model is string identifier corresponding to the model to use. Options are "RIVER", "Watershed_exact", and "Watershed_approximate"

"Watershed_exact" is Watershed where parameters are optimized via exact inference (tractable and recommended when E is small. A general rule of thumb is if E is less than equal to 4, exact inference should be used). 
"Watershed_approximate" is Watershed where parameters are optimized using approximate inference. This approach is tractable when E is large. For example, we used this to model the related outlier signals from 49 tissues (see bioRxiv preprint).

There are several other command line arguments that can be specified. A description can be found in "evaluate_watershed.R"

## evaluate_watershed.R deliverables

Running evaluate_watershed.R will save an R object to $output_prefix'evaluation_object.rds'.
This object can be accesed in R with the following command:
evaluation_object <- readRDS(evaluation_object_file_name)

* evaluation_object$model_params contains all of the optimized Watershed parameters.
* evaluation_object$gam_model_params contains all optimized GAM parameters
* evaluation_object$auc contains information on evaluation (precion-recall stats, etc)
* evaluation_object$auc[[1]]$evaROC$watershed_pr_auc gives the area under the precision recall curve (for Watershed/RIVER) in the first outlier dimension (base 1)
* evaluation_object$auc[[3]]$evaROC$watershed_pr_auc gives the area under the precision recall curve (for Watershed/RIVER) in the third outlier dimension (base 1)
* evaluation_object$auc[[1]]$evaROC$GAM_pr_auc gives the area under the precision recall curve (for GAM) in the first outlier dimension (base 1)
* evaluation_object$auc[[1]]$evaROC$watershed_precision gives a vector of precision values (for Watershed/RIVER) in the first outlier dimension (base 1). The vector spans posterior thresholds generated by the R library "PRROC"
* evaluation_object$auc[[1]]$evaROC$watershed_recall gives a vector of recall values (for Watershed/RIVER) in the first outlier dimension (base 1). This vector is the same length as precision vector above (and can be therefor used together to make a PR-curve). The vector spans posterior thresholds generated by the R library "PRROC"
* evaluation_object$auc[[1]]$evaROC$GAM_precision gives a vector of precision values (for GAM) in the first outlier dimension (base 1). The vector spans posterior thresholds generated by the R library "PRROC"
* evaluation_object$auc[[1]]$evaROC$GAM_recall gives a vector of recall values (for GAM) in the first outlier dimension (base 1). This vector is the same length as precision vector above (and can be therefor used together to make a PR-curve). The vector spans posterior thresholds generated by the R library "PRROC"

## Running predict_watershed.R
TO DO

## predict_watershed.R deliverables
TO DO




## Authors

* **Ben Strober** -- [BennyStrobes](https://github.com/BennyStrobes) -- bstrober3@gmail.com